app/controllers/email_confirmations_controller.rb

```diff
@@ -17,10 +19,37 @@ def mfa_update
     if @user.mfa_enabled? && @user.otp_verified?(params[:otp])
       confirm_email
     else
-      @form_url       = mfa_update_email_confirmations_url(token: @user.confirmation_token)
-      flash.now.alert = t("multifactor_auths.incorrect_otp")
-      render template: "multifactor_auths/otp_prompt", status: :unauthorized
+      @form_mfa_url = mfa_update_email_confirmations_url(token: @user.confirmation_token)
+      login_failure(t("multifactor_auths.incorrect_otp"))
+    end
+  end
+
+  def webauthn_update
+    @user = User.find(session.dig(:webauthn_authentication, "user"))
+    @challenge = session.dig(:webauthn_authentication, "challenge")
+
+    if params[:credentials].blank?
+      login_failure("Credentials required")
+      return
     end
+
+    @credential = WebAuthn::Credential.from_get(params[:credentials])
+
+    @webauthn_credential = @user.webauthn_credentials.find_by(
+      external_id: @credential.id
+    )
+
+    @credential.verify(
+      @challenge,
+      public_key: @webauthn_credential.public_key,
+      sign_count: @webauthn_credential.sign_count
+    )
+
+    @webauthn_credential.update!(sign_count: @credential.sign_count)
+
+    confirm_email
```

Comment:
It looks like `confirm_email` does a redirect to root, and the notice doesn't show up, because of an issue with the JS (there's another render that erases the flash) https://github.com/rubygems/rubygems.org/pull/2865#discussion_r1035366844

What happens if you try `flash.keep` in confirm_email? Does the flash message show up and when does it disappear? 

Either way, I'll create an issue should be created to revisit fixing this issue for the long term.

Reply:
Using `flash.keep` unfortunately doesn't help 😢

Reply:
@ericherscovich and I discussed this and decided that we'll punt this issue to be fixed as a part of https://github.com/rubygems/rubygems.org/issues/3372.